doc/developer/design/20210728_error_handling_surfacing_connector_failures.md

```diff
@@ -0,0 +1,194 @@
+understand the current state.
+
+### Current State
+
+#### Where/when do we interact with external systems?
+
+- When creating a source/sink:
+  - (coord) purification
+- When creating a source/sink or when materialized is restarted:
+  - (coord) initial connector setup, for example:
+  - (coord) reading Kafka consistency topic
+  - (coord) publishing schemas
+  - (coord) creating topics
+  - (coord) creating output file, checking input file
+- Continuously:
+  - (coord) metadata loops, for example fetching new Kafka partitions, listening
+   on BYO topics, listening on SQS for S3 updates
+  - (dataflow) actual data ingest and writing
+
+#### Failures during purification
+
+In general, failures during purification lead to the statement being aborted,
+nothing gets added to the catalog. This seems fair, because we cannot store
+un-purified statements in the catalog and purification happens infrequently,
+only when a statement is issued.
+
+Most purification logic doesn't specify timeouts, only S3 and Kinesis have a 1
+second timeout for the validation call. I believe none of the purification
+checks do retries
+
+#### Failures During Source/Sink Creation
+
+The behaviour is very similar (the same?) to failures during purification:
+nothing gets added to the catalog when there are failures.
+
+Some calls have custom timeouts, some don't. I believe none of the calls have
+retries.
+
+#### Failures During "Normal Operations"
+
+Failures during execution don't bring down materialized. If failures happen,
+they are only surfaced in the log. Materialized doesn't "know" about the state
+of a connector.
+
+There are no global config options for timeouts/retries/back-off interval.
+
+#### Failures During Restart (restarting a materialized with things in the catalog)
+
+Failures during sink (re-)creation kill materialized when restarting with a
+filled catalog.  Failures during source creation are fine.
+
+The reason is that (re-)creating a source runs no source-specific code in the
+coordinator. Everything happens in dataflow, where we have retries. This is not
+true for sinks.
+
+### Proposed Changes
+
+The reaction to failures at various stages should uniform across different
+connectors and informed by these rules:
+
+ - Failures during connector creation (which includes purification) immediately
+   report an error back to the user. No catalog entry is created.
+ - Failures during runtime need to be surfaced to the user in system tables.
+ - Failures that occur during a restart with an already filled catalog must not
+   bring down Materialize. Instead, errors that occur must be reported to the
+   user for each individual connector.
+
+Additionally, we should have global settings that control timeouts, retries,
+and the like.
+
+The concrete steps to achieve this, in the order we should address them:
+
+ - Introduce global settings for `timeout`, `num retries`, and potentially
```

Comment:
Are these user-settable globals? Or just a default bundle of `Retry` options that get plumbed around but are hardcoded into the binary? I think I like the second thing better! Retry settings to me feel like something that the user typically doesn't have much useful insight into, at least not at the global level. If you do have an opinion on retry configuration it's usually in the context of one specific retry loop that is misconfigured.

Reply:
I initially intended it to be user-settable globals, but then thought along the same lines as you and added some text under Alternatives that takes that back. I'll remove this from here.

By now, I'm not even sure we need global `Retry` options except maybe for restarting whole connectors/views.